packages/plugin-hardhat/src/platform/propose-upgrade.ts

```diff
@@ -0,0 +1,113 @@
+import '@openzeppelin/hardhat-upgrades/dist/type-extensions';
+import {
+  getImplementationAddress,
+  isBeaconProxy,
+  isTransparentOrUUPSProxy,
+  isTransparentProxy,
+} from '@openzeppelin/upgrades-core';
+import { ProposalResponse } from 'defender-admin-client';
+import { ContractFactory, ethers } from 'ethers';
+import { FormatTypes } from 'ethers/lib/utils';
+import { HardhatRuntimeEnvironment } from 'hardhat/types';
+import { PlatformSupportedOptions, UpgradeOptions } from '../utils';
+import { getNetwork, getAdminClient, setPlatformDefaults } from './utils';
+import type { VerificationResponse } from './verify-deployment';
+
+export interface ExtendedProposalResponse extends ProposalResponse {
+  txResponse?: ethers.providers.TransactionResponse;
+  verificationResponse?: VerificationResponse;
+}
+
+export type ProposeUpgradeFunction = (
+  proxyAddress: string,
+  contractNameOrImplFactory: string | ContractFactory,
+  opts?: ProposalOptions,
+) => Promise<ExtendedProposalResponse>;
+
+export interface ProposalOptions extends UpgradeOptions, PlatformSupportedOptions {
+  title?: string;
+  description?: string;
+  proxyAdmin?: string;
+  multisig?: string;
+  multisigType?: 'Gnosis Safe' | 'Gnosis Multisig' | 'EOA';
+  bytecodeVerificationReferenceUrl?: string;
+}
+
+export function makeProposeUpgrade(hre: HardhatRuntimeEnvironment, platformModule: boolean): ProposeUpgradeFunction {
+  return async function proposeUpgrade(proxyAddress, contractNameOrImplFactory, opts = {}) {
+    setPlatformDefaults(hre, platformModule, opts);
+
+    const client = getAdminClient(hre);
+    const network = await getNetwork(hre);
+
+    const { title, description, proxyAdmin, multisig, multisigType, ...moreOpts } = opts;
+
+    if (await isBeaconProxy(hre.network.provider, proxyAddress)) {
+      throw new Error(`Beacon proxy is not currently supported with defender.proposeUpgrade()`);
+    } else if (
+      !multisig &&
+      (await isTransparentOrUUPSProxy(hre.network.provider, proxyAddress)) &&
+      !(await isTransparentProxy(hre.network.provider, proxyAddress))
+    ) {
+      throw new Error(`Multisig address is a required property for UUPS proxies`);
+    } else {
+      // try getting the implementation address so that it will give an error if it's not a transparent/uups proxy
+      await getImplementationAddress(hre.network.provider, proxyAddress);
+    }
+
+    const implFactory =
+      typeof contractNameOrImplFactory === 'string'
+        ? await hre.ethers.getContractFactory(contractNameOrImplFactory)
+        : contractNameOrImplFactory;
+    const contractName = typeof contractNameOrImplFactory === 'string' ? contractNameOrImplFactory : undefined;
```

Comment:
So if you want verification you need to provide the contract name as opposed to the impl factory?

Reply:
Removed this since verification is handled automatically by Deploy API